webapp/src/main/java/com/box/l10n/mojito/service/thirdparty/ThirdPartyTMSSmartlingGlossary.java

```diff
@@ -0,0 +1,300 @@
+                virtualAssetService.deleteTextUnit(asset.getId(), assetTextUnit.getName());
+            }
+        }
+    }
+
+    private String downloadGlossaryFile(String glossaryId) {
+        return Mono.fromCallable(() -> smartlingClient.downloadGlossaryFile(accountId, glossaryId)).retryWhen(getRetryConfiguration()
+                .doBeforeRetry(e -> logger.info(String.format("Retrying after failure to download glossary file for glossary id: %s", glossaryId), e.failure())))
+                .doOnError(e -> {
+                    String msg = String.format("Error downloading glossary file from Smartling for glossary id: %s", glossaryId);
+                    logger.error(msg, e);
+                    throw new SmartlingClientException(msg, e);
+                }).blockOptional()
+                .orElseThrow(() -> new SmartlingClientException(String.format("Error downloading glossary file from Smartling for glossary id: %s, optional is empty", glossaryId)));
+    }
+
+    private String getGlossaryName(String glossaryUID) {
+        return Mono.fromCallable(() -> smartlingClient.getGlossaryDetails(accountId, glossaryUID).getName())
+                .retryWhen(getRetryConfiguration()
+                        .doBeforeRetry(e -> logger.info(String.format("Retrying after failure to retrieve glossary details from Smartling for glossary id: %s", glossaryUID), e.failure())))
+                .doOnError(e -> {
+                    String msg = String.format("Error retrieving glossary details from Smartling for glossary id: %s", glossaryUID);
+                    logger.info(msg, e);
+                    throw new SmartlingClientException(msg, e);
+                }).blockOptional()
+                .orElseThrow(() -> new SmartlingClientException(String.format("Error retrieving glossary details from Smartling for glossary id: %s, optional is empty", glossaryUID)));
+    }
+
+    private String getRepositorySourceLocale(Repository repository) {
+        return repository.getRepositoryLocales().stream()
+                .filter(repositoryLocale -> repositoryLocale.getParentLocale() == null)
+                .map(repositoryLocale -> repositoryLocale.getLocale().getBcp47Tag())
+                .findFirst()
+                .orElseThrow(() -> new ThirdPartyTMSGlossarySyncException("Unable to find source locale for repository: " + repository.getName()));
+    }
+
+    private VirtualAsset getVirtualAsset(Repository repository, String glossaryName) {
+        VirtualAsset virtualAsset = new VirtualAsset();
+        virtualAsset.setRepositoryId(repository.getId());
+        virtualAsset.setPath(glossaryName);
+        virtualAsset.setDeleted(false);
+        try {
+            virtualAsset = virtualAssetService.createOrUpdateVirtualAsset(virtualAsset);
+        } catch (VirtualAssetBadRequestException e) {
+            throw new ThirdPartyTMSGlossarySyncException(e.getMessage(), e);
```

Comment:
what would be the error here?

Reply:
Errors are around invalid repository id e.g. "A repository id must be provided", "A valid repository must be provided when creating an asset"